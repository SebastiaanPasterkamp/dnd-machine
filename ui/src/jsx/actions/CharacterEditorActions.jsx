import React from 'react';
import Reflux from 'reflux';
import _ from 'lodash';

const CharacterEditorActions = Reflux.createActions({
    "editCharacter": {asyncResult: true},
    "resetCharacter": {asyncResult: true},
    "postCharacter": {asyncResult: true},
    "patchCharacter": {asyncResult: true},
    "addChange": {},
    "removeChange": {},
    "addAbilityScoreIncrease": {},
    "removeAbilityScoreIncrease": {},
});

function getCharacter(action, id) {
    if (id === null) {
        action.completed({});
        return;
    }

    fetch(`/character/api/${ id }`, {
        credentials: 'same-origin',
        method: 'GET',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest'
        }
    })
    .then(response => response.json())
    .then(result => action.completed(result))
    .catch(error => {
        console.log(error);
        action.failed(id, error);
    });
};

CharacterEditorActions.editCharacter.listen(
    (id=null) => getCharacter(
        CharacterEditorActions.editCharacter,
        id,
    )
);

CharacterEditorActions.resetCharacter.listen(
    (id=null) => getCharacter(
        CharacterEditorActions.resetCharacter,
        id,
    )
);

const saveCharacter = (
    action, method, data, id=null, callback=null
) => {
    const path = '/' + _.filter(
        ['character', 'api', id]
    ).join('/');

    fetch(path, {
        credentials: 'same-origin',
        method,
        'headers': {
            'X-Requested-With': 'XMLHttpRequest',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(data)
    })
    .then((response) => response.json())
    .then((result) => {
        action.completed(result.id, result, callback);
    })
    .catch((error) => {
        action.failed(id, error);
    });
};

CharacterEditorActions.postCharacter.listen(
    (data, id=null, callback=null) => saveCharacter(
        CharacterEditorActions.postCharacter,
        'POST',
        data,
        id,
        callback,
    )
);

CharacterEditorActions.patchCharacter.listen(
    (data, id=null, callback=null) => saveCharacter(
        CharacterEditorActions.patchCharacter,
        'PATCH',
        data,
        id,
        callback,
    )
);

export default CharacterEditorActions;
