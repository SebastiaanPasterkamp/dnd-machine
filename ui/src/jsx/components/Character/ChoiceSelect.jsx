import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import LazyComponent from '../LazyComponent.jsx';
import TabComponent from '../TabComponent.jsx';

import CharacterConfig from './CharacterConfig.jsx';

class ChoiceSelect extends LazyComponent
{
    constructor(props) {
        super(props);
    }

    render() {
        return <TabComponent
                tabConfig={this.props.options}
                >
            {_.map(this.props.options, (option, index) => {
                const props = {
                    index: this.props.index.concat([index]),
                    getCurrent: this.props.getCurrent,
                    getItems: this.props.getItems,
                    onChange: this.props.onChange,
                    config: option.config
                };

                return <div key={index}>
                    {option.description
                        ? <MDReactComponent
                            text={option.description}
                            />
                        : null
                    }
                    <CharacterConfig
                        {...props}
                        />
                </div>;
            })}
        </TabComponent>;
    }
};

ChoiceSelect.propTypes = {
    onChange: PropTypes.func.isRequired,
    getCurrent: PropTypes.func.isRequired,
    getItems: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    description: PropTypes.string,
};

export default ChoiceSelect;
