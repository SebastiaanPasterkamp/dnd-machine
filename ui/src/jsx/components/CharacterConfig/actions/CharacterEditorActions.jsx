import React from 'react';
import Reflux from 'reflux';
import {
    filter,
} from 'lodash/fp';

const CharacterEditorActions = Reflux.createActions({
    "editCharacter": {asyncResult: true},
    "resetCharacter": {asyncResult: true},
    "postCharacter": {asyncResult: true},
    "patchCharacter": {asyncResult: true},
    "addChange": {},
    "removeChange": {},
});

function getCharacter(action, id) {
    if (id === null) {
        action.completed({});
        return;
    }

    fetch(`/character/api/${ id }`, {
        credentials: 'same-origin',
        method: 'GET',
        'headers': {
            'Accept': 'application/json'
        }
    })
    .then(response => response.json())
    .then(result => action.completed(result))
    .catch(error => {
        console.log(error);
        action.failed(id, error);
    });
};

CharacterEditorActions.editCharacter.listen(
    (id=null) => getCharacter(
        CharacterEditorActions.editCharacter,
        id,
    )
);

CharacterEditorActions.resetCharacter.listen(
    (id=null) => getCharacter(
        CharacterEditorActions.resetCharacter,
        id,
    )
);

const saveCharacter = (
    action, method, data, id=null, callback=null
) => {
    const path = '/' + filter(null, [ 'character', 'api', id ]).join('/');

    fetch(path, {
        credentials: 'same-origin',
        method,
        'headers': {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(data)
    })
    .then((response) => response.json())
    .then((result) => {
        action.completed(result.id, result, callback);
    })
    .catch((error) => {
        action.failed(id, error);
    });
};

CharacterEditorActions.postCharacter.listen(
    (data, id=null, callback=null) => saveCharacter(
        CharacterEditorActions.postCharacter,
        'POST',
        data,
        id,
        callback,
    )
);

CharacterEditorActions.patchCharacter.listen(
    (data, id=null, callback=null) => saveCharacter(
        CharacterEditorActions.patchCharacter,
        'PATCH',
        data,
        id,
        callback,
    )
);

export default CharacterEditorActions;
