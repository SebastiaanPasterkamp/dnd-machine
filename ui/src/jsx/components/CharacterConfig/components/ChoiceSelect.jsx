import React from 'react';
import PropTypes from 'prop-types';
import {
    entries,
    filter,
    flow,
    map,
} from 'lodash/fp';

import LazyComponent from '../../LazyComponent.jsx';
import TabComponent from '../../TabComponent.jsx';

import CharacterConfig from '../CharacterConfig.jsx';
import MatchesFilters from '../utils/MatchesFilters';

export const ChoiceSelect = function({ options, filter: filters }) {
    const filtered = filter(
        (option) => !option.hidden && MatchesFilters(option, filters)
    )(options);

    return (
        <TabComponent tabConfig={filtered}>
            {flow(entries, map(
                ([index, option]) => (
                    <CharacterConfig
                        key={ index }
                        config={ [option] }
                    />
                )
            ))(filtered)}
        </TabComponent>
    );
};

ChoiceSelect.propTypes = {
    type: PropTypes.oneOf(['choice']).isRequired,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            description: PropTypes.string,
        })
    ).isRequired,
    description: PropTypes.string,
};

export default ChoiceSelect;
