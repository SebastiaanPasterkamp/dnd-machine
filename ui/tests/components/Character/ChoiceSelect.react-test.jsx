import React from 'react';
import { mount } from 'enzyme';

import ChoiceSelect from 'components/Character/ChoiceSelect.jsx';

import CharacterEditorActions from 'actions/CharacterEditorActions.jsx';

const props = {
    type: 'choice',
    description: "One choice please",
    options: [{
        label: 'Choice A',
        type: 'ability_score',
        limit: 1,
    }, {
        label: 'Choice B',
        description: 'Some *text* here',
        type: 'ability_score',
        limit: 2,
    }],
};

describe('Component: ChoiceSelect', () => {
    it('should render with minimum props', () => {
        const wrapper = mount(
            <ChoiceSelect
                type="choice"
                options={[]}
                />
        );

        expect(wrapper)
            .toMatchSnapshot();
    });

    it('should render with all props', () => {
        const wrapper = mount(
            <ChoiceSelect
                {...props}
                />
        );

        expect(wrapper)
            .toMatchSnapshot();
    });

    it('should emit changes while switching tabs', () => {
        const add = jest.spyOn(
            CharacterEditorActions,
            'addAbilityScoreIncrease'
        );
        const remove = jest.spyOn(
            CharacterEditorActions,
            'removeAbilityScoreIncrease'
        );
        const wrapper = mount(
            <ChoiceSelect
                {...props}
                />
        );

        expect(add)
            .toBeCalledWith(1);

        add.mockClear();

        wrapper
            .find('.tab-component__tab a')
            .at(1)
            .simulate('click');

        expect(remove)
            .toBeCalledWith(1);
        expect(add)
            .toBeCalledWith(2);
    });
});
