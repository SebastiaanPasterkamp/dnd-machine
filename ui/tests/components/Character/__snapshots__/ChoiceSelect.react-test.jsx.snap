// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ChoiceSelect should render with all props 1`] = `
<ChoiceSelect
  description="One choice please"
  options={
    Array [
      Object {
        "label": "Choice A",
        "limit": 1,
        "type": "ability_score",
      },
      Object {
        "description": "Some *text* here",
        "label": "Choice B",
        "limit": 2,
        "type": "ability_score",
      },
    ]
  }
  type="choice"
>
  <TabComponent
    tabConfig={
      Array [
        Object {
          "label": "Choice A",
          "limit": 1,
          "type": "ability_score",
        },
        Object {
          "description": "Some *text* here",
          "label": "Choice B",
          "limit": 2,
          "type": "ability_score",
        },
      ]
    }
  >
    <div
      className="nice-tabs-wrapper tab-component"
    >
      <ul
        className="nice-tabs bold"
      >
        <li
          className="tab-component__tab current"
          key="tab-0/.$0"
        >
          <a
            className="cursor-pointer"
            onClick={null}
          >
            Choice A
          </a>
        </li>
        <li
          className="tab-component__tab"
          key="tab-1/.$1"
        >
          <a
            className="cursor-pointer"
            onClick={[Function]}
          >
            Choice B
          </a>
        </li>
      </ul>
      <ul
        className="nice-tab-content"
      >
        <li
          className="current"
          key=".$0"
        >
          <CharacterConfig
            config={
              Array [
                Object {
                  "label": "Choice A",
                  "limit": 1,
                  "type": "ability_score",
                },
              ]
            }
            key="0"
          >
            <FormGroup
              key="0"
              label="Choice A"
            >
              <div
                className="nice-form-group"
              >
                <label>
                  Choice A
                </label>
                <AbilityScoreSelect
                  limit={1}
                  type="ability_score"
                />
              </div>
            </FormGroup>
          </CharacterConfig>
        </li>
      </ul>
    </div>
  </TabComponent>
</ChoiceSelect>
`;

exports[`Component: ChoiceSelect should render with minimum props 1`] = `
<ChoiceSelect
  options={Array []}
  type="choice"
>
  <TabComponent
    tabConfig={Array []}
  >
    <div
      className="nice-tabs-wrapper tab-component"
    >
      <ul
        className="nice-tabs bold"
      />
      <ul
        className="nice-tab-content"
      />
    </div>
  </TabComponent>
</ChoiceSelect>
`;
